tensorpipe/transport/ibv/reactor.cc

```diff
@@ -17,8 +17,32 @@ namespace ibv {
 
 Reactor::Reactor(IbvLib ibvLib, IbvDeviceList deviceList)
     : ibvLib_(std::move(ibvLib)) {
+  bool found = false;
   TP_DCHECK_GE(deviceList.size(), 1);
-  ctx_ = createIbvContext(getIbvLib(), deviceList[0]);
+
+  // If the deviceList contains multiple ibv devices, we will select the
+  // device of the port whose port_state is active, instead of just selecting
+  // the first device in the deviceList by default.
+  for (int i = 0; i < deviceList.size(); i++) {
+    IbvContext tp_ctx_;
+    IbvLib::port_attr portAttr;
+    std::memset(&portAttr, 0, sizeof(portAttr));
+    tp_ctx_ = createIbvContext(getIbvLib(), deviceList[i]);
+    TP_CHECK_IBV_INT(ibvLib.query_port(tp_ctx_.get(), kPortNum, &portAttr));
+    if (portAttr.state == IbvLib::port_state::PORT_ACTIVE) {
+      ctx_ = std::move(tp_ctx_);
+      found = true;
+      break;
+    } else {
+      TP_VLOG(8) << "IbvDevice " << deviceList[i].name << " port "
+                 << unsigned(kPortNum) << " state is "
+                 << ibvLib.port_state_str(portAttr.state)
+                 << " , so skip this device";
+    }
+  }
+
+  TP_THROW_ASSERT_IF(found == false) << "Unable to find available ibv device";
```

Comment:
If we can't find any usable devices we shouldn't consider it an error (and crash the program), instead we should just disable the ibv transport. The logic to do so happens here: https://github.com/pytorch/tensorpipe/blob/bb1473a4b38b18268e8693044afdb8635bc8351b/tensorpipe/transport/ibv/context_impl.cc#L58-L62

Could you move your code to that file? You will probably need to change the constructor of the Reactor class so that it takes a IbvContext object, instead of an IbvDeviceList.